modules/core/src/main/java/org/apache/ignite/internal/util/VersatileReadWriteLock.java

```diff
@@ -0,0 +1,498 @@
+                if (trySwitchStateToWriteLocked()) {
+                    return true;
+                }
+
+                Thread.sleep(SLEEP_MILLIS);
+
+                if (System.nanoTime() - startNanos >= timeoutNanos) {
+                    return false;
+                }
+            }
+        } finally {
+            decrementPendingWriteLocks();
+        }
+    }
+
+    /**
+     * Releases the write lock.
+     *
+     * @throws IllegalMonitorStateException thrown if the write lock is not acquired.
+     */
+    public void writeUnlock() {
+        int curState = state;
+        // There could still be some read locks if the write lock was taken forcefully.
+        int readLocks = readLocks(curState);
+
+        if (!writeLocked(curState)) {
+            throw new IllegalMonitorStateException();
+        }
+
+        boolean b = compareAndSet(STATE_VH, state, state(false, readLocks));
+
+        assert b;
+
+        notifyWriteLockSolicitors();
+        notifyReadLockSolicitors();
+    }
+
+    private void notifyWriteLockSolicitors() {
+        if (writeLockSolicitors.isEmpty()) {
+            return;
+        }
+
+        for (Iterator<CompletableFuture<Void>> iterator = writeLockSolicitors.iterator(); iterator.hasNext(); ) {
+            CompletableFuture<Void> future = iterator.next();
+
+            if (!tryWriteLock()) {
+                // Someone has already acquired an impeding lock, we're too late, let's wait for next opportunity.
+                break;
+            }
+
+            decrementPendingWriteLocks();
+
+            asyncContinuationExecutor.execute(() -> future.complete(null));
+
+            iterator.remove();
+        }
+    }
+
+    private void notifyReadLockSolicitors() {
+        if (readLockSolicitors.isEmpty()) {
+            return;
+        }
+
+        for (Iterator<CompletableFuture<Void>> iterator = readLockSolicitors.iterator(); iterator.hasNext(); ) {
+            CompletableFuture<Void> future = iterator.next();
+
+            if (!tryReadLock()) {
+                // Someone has already acquired a write lock, we're too late, let's wait for next opportunity.
+                break;
+            }
+
+            asyncContinuationExecutor.execute(() -> {
+                if (!future.complete(null)) {
+                    // The one who added this future has already taken the read lock after adding the future; as they have completed the
+                    // future, this is us who needs to unlock the excess.
+                    readUnlock();
+                }
+            });
+
+            iterator.remove();
+        }
+    }
+
+    /**
+     * Returns {@code true} on success.
+     *
+     * @param varHandle VarHandle.
+     * @param expect    Expected.
+     * @param update    Update.
+     * @return {@code True} on success.
+     */
+    private boolean compareAndSet(VarHandle varHandle, int expect, int update) {
+        return varHandle.compareAndSet(this, expect, update);
+    }
```

Comment:
Why did you do this? Now you have to rely on inline if you want Java to optimize your VH. Direct access to static final constants is preferable, this way JIT knows what to optimize without additional steps

Reply:
Good catch. Inlined the methods